Controller argspec

bcoca opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

bcoca committed:
controller side argspec

  read docs to create argspec (allows for handling non python modules)
  no_log? add property for marking options as such
  restrictions? handle the required_if/etc functions
  attributes? handles check/diff mode and more

bcoca committed:
nolog#

bcoca committed:
move so modules can use

bcoca committed:
refining

bcoca committed:
move module finding to it's own function

bcoca committed:
added TODO to the meat of it

felixfontein commented:
CC @russoz I'm pretty sure you'll be interested in this :)

russoz commented:
Oh yeah! Thanks for the heads up! 😁

tremble commented on lib/ansible/module_utils/parsing/docs_to_spec.py:
```diff
@@ -0,0 +1,148 @@
+# Copyright: 2017, Ansible Project
```
```suggestion
# Copyright: 2017, Ansible Project
```

It's not 2017 any more...  Not sure if "Ansible Project" is the correct assignee anymore either.

bcoca committed:
add validation parts

bcoca committed:
update cright

bcoca committed:
fix deprecation passthrough

mattclay commented on lib/ansible/module_utils/parsing/docs_to_spec.py:
```diff
@@ -0,0 +1,166 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from collections.abc import Sequence
```
```suggestion
from ansible.module_utils.common._collections_compat import Sequence
```

bcoca committed:
Update lib/ansible/module_utils/parsing/docs_to_spec.py

Co-authored-by: Matt Clay <matt@mystile.com>

bcoca committed:
todo or not todo

bcoca committed:
avoid noise

bcoca commented:
cc @briantist

bcoca committed:
fix imporst

bcoca committed:
clog

bcoca committed:
s

bcoca committed:
fix invocation

russoz commented on lib/ansible/module_utils/parsing/docs_to_spec.py:
```diff
@@ -0,0 +1,163 @@
+            name = 'required_if'
+            rest = [r['if'], r['equals'], r['required']]
+    else:
+        for ding in ('exclusive', 'toghether', 'one_of'):
```
typo
```suggestion
        for ding in ('exclusive', 'together', 'one_of'):
```

russoz commented on lib/ansible/module_utils/parsing/docs_to_spec.py:
```diff
@@ -0,0 +1,163 @@
+   - description: You cannot use 'c' and 'x' at the same time
+     exclusive: c, x
+
+    # required_toghether
+   - description: 'a' and 'b' required toghether
+     toghether: a, b
```
typo
```suggestion
    # required_together
   - description: 'a' and 'b' required together
     together: a, b
```

russoz reviewed (commented):
Hi @bcoca I know it is still draft, just picked up a typo. Cheers

bcoca committed:
Apply suggestions from code review

Co-authored-by: Alexei Znamensky <103110+russoz@users.noreply.github.com>

bcoca committed:
fix to standards

bcoca committed:
relocate docs

bcoca committed:
st it

bcoca committed:
sldfk

bcoca committed:
setup generic invocation

todo:
   - needs to handle no_log, once it has argspec
   - update test to user friendly message
   - update module_utils/basic.cs to avoid creating by default and adjust msg

bcoca committed:
lint

bcoca committed:
typed
